HandWriter.py

```diff
@@ -0,0 +1,269 @@
+        if self.doc_path == '':
+            self.sleep_btn_write()
+            return
+        try:
+            self.document = Document(self.doc_path)
+        except PackageNotFoundError:
+            self.sleep_btn_write()
+            return
+        self.wake_btn_write()
+        
+    # Parse document on a thread separate from main UI thread
+    def parse_document(self):
+        self.progress.setRange(0, 0)
+        self.progress.setStyleSheet(self.stylesheet_busy_progressbar)
+        self.start_parsing()
+    # Uncomment to help with debugging:
+#        if self.thread.isRunning():
+#            print('Thread started')
+
+    def start_parsing(self):
+        self.thread = ParserThread(self.doc_path, self.document)
+        self.thread.change_value.connect(self.stop_progressbar)
+        self.thread.key_exception.connect(self.popup_keyerror)
+        self.thread.start()        
+        
+    def stop_progressbar(self):
+        self.thread.requestInterruption()
+        self.sleep_btn_write()
+        # Uncomment to help with debugging:
+#        if self.thread.isFinished():
+#            print("Thread terminated")
+        self.progress.setRange(0, 1)
+        self.progress.setStyleSheet(self.stylesheet_complete_progressbar)
+        self.progress.setTextVisible(False)
+        self.progress.setValue(1)
+    
+    def popup_keyerror(self, foreign_char):
+        print(foreign_char)
+        self.stop_progressbar()
+        error_popup = QtWidgets.QMessageBox(self.centralwidget)
+        error_popup.setIcon(QtWidgets.QMessageBox.Critical)
+        error_popup.setWindowTitle('Error: Unable to write character')
+        error_popup.setText('The character ' + foreign_char + ' has not been fed into this version of HandWriter. Raise an issue on the official GDGVIT repo')
+        error_popup.setStandardButtons(QtWidgets.QMessageBox.Ok)
+        error_popup.show()
+        
+    def wake_btn_write(self):
+        self.btn_write.setEnabled(True)
+        self.btn_write.setStyleSheet(self.stylesheet_write_active)
+        self.btn_write.setFont(self.font_awake)
+
+    def sleep_btn_write(self):
+        self.btn_write.setEnabled(False)
+        self.btn_write.setStyleSheet(self.stylesheet_write_inactive)
+        self.btn_write.setFont(self.font_asleep)
+
+    def check_stopping_frame(self):
+        if self.logo_movie.currentFrameNumber() == 110: # 110 - Frame where DSC logo completes
+            self.logo_movie.stop()
+
+# Thread class for executing application logic separate from main UI thread            
+class ParserThread(QtCore.QThread):
+    def __init__(self, doc_path, document):
+        super(ParserThread, self).__init__()
+        self.doc_path = doc_path
+        self.document = document
+
+    change_value = QtCore.pyqtSignal()
+    key_exception = QtCore.pyqtSignal(str)
+    
+    def run(self):
+        CHARS_PER_LINE = 54
+        LINES_PER_PAGE = 30
+        with open('hashes.pickle', 'rb') as f:
```

Comment:
use this file name as a constant. Define it somewhere so that you can easily change this without editing it multiple places.

Reply:
Done